File: 2024/day08/node/solution.js
import { readFileSync } from 'fs';

/**
 * Create a string key for a coordinate pair.
 * @param {number} r - Row index
 * @param {number} c - Column index
 * @returns {string} Coordinate key in format "r,c"
 */
const coordKey = (r, c) => `${r},${c}`;

/**
 * Parse input file into grid dimensions and antenna positions grouped by frequency.
 * @param {string} filename - Path to input file
 * @returns {{rows: number, cols: number, antennas: Map<string, Array<[number, number]>>}}
 */
function parseInput(filename) {
    const grid = readFileSync(filename, 'utf-8').trim().split('\n');
    const rows = grid.length;
    const cols = grid[0]?.length ?? 0;

    // Group antenna positions by frequency
    const antennas = new Map();
    grid.forEach((row, r) => {
        [...row].forEach((ch, c) => {
            if (ch !== '.') {
                if (!antennas.has(ch)) {
                    antennas.set(ch, []);
                }
                antennas.get(ch).push([r, c]);
            }
        });
    });

    return { rows, cols, antennas };
}

/**
 * Calculate antinode positions for Part 1.
 * For each antenna pair, calculate two antinodes:
 * - One beyond antenna 1 at position 2*A1 - A2
 * - One beyond antenna 2 at position 2*A2 - A1
 * @returns {number} Count of unique antinode positions
 */
function part1() {
    const { rows, cols, antennas } = parseInput('../input.txt');
    const antinodes = new Set();

    for (const positions of antennas.values()) {
        // For each pair of antennas with same frequency
        for (let i = 0; i < positions.length; i++) {
            for (let j = i + 1; j < positions.length; j++) {
                const [r1, c1] = positions[i];
                const [r2, c2] = positions[j];

                // Calculate the two antinodes (2:1 distance ratio)
                // Antinode beyond antenna 1 (away from antenna 2)
                const ar1 = 2 * r1 - r2;
                const ac1 = 2 * c1 - c2;
                // Antinode beyond antenna 2 (away from antenna 1)
                const ar2 = 2 * r2 - r1;
                const ac2 = 2 * c2 - c1;

                // Add if within bounds
                if (ar1 >= 0 && ar1 < rows && ac1 >= 0 && ac1 < cols) {
                    antinodes.add(coordKey(ar1, ac1));
                }
                if (ar2 >= 0 && ar2 < rows && ac2 >= 0 && ac2 < cols) {
                    antinodes.add(coordKey(ar2, ac2));
                }
            }
        }
    }

    return antinodes.size;
}

/**
 * Calculate antinode positions for Part 2.
 * For each antenna pair, extend along the line in both directions,
 * marking all grid positions including the antennas themselves.
 * This accounts for resonant harmonics at any integer multiple distance.
 * @returns {number} Count of unique antinode positions
 */
function part2() {
    const { rows, cols, antennas } = parseInput('../input.txt');
    const antinodes = new Set();

    for (const positions of antennas.values()) {
        // For each pair of antennas with same frequency
        for (let i = 0; i < positions.length; i++) {
            for (let j = i + 1; j < positions.length; j++) {
                const [r1, c1] = positions[i];
                const [r2, c2] = positions[j];
                const dr = r2 - r1;
                const dc = c2 - c1;

                // Extend in both directions along the line
                // Direction 1: from antenna 1 towards and beyond antenna 2
                let r = r1, c = c1;
                while (r >= 0 && r < rows && c >= 0 && c < cols) {
                    antinodes.add(coordKey(r, c));
                    r += dr;
                    c += dc;
                }

                // Direction 2: from antenna 1 away from antenna 2
                r = r1 - dr;
                c = c1 - dc;
                while (r >= 0 && r < rows && c >= 0 && c < cols) {
                    antinodes.add(coordKey(r, c));
                    r -= dr;
                    c -= dc;
                }
            }
        }
    }

    return antinodes.size;
}

console.log('Part 1:', part1());
console.log('Part 2:', part2());

File: 2024/day08/java/Solution.java
import java.io.*;
import java.util.*;

public class Solution {
    record Point(int r, int c) {}

    record Input(int rows, int cols, Map<Character, List<Point>> antennas) {}

    private static boolean inBounds(int r, int c, int rows, int cols) {
        return r >= 0 && r < rows && c >= 0 && c < cols;
    }

    /**
     * Parses the input file and extracts grid dimensions and antenna positions.
     *
     * @param filename the path to the input file
     * @return an Input record containing grid dimensions and antenna positions grouped by frequency
     * @throws IOException if the file cannot be read
     */
    static Input parseInput(String filename) throws IOException {
        var grid = new ArrayList<String>();

        try (var br = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = br.readLine()) != null) {
                grid.add(line);
            }
        }

        int rows = grid.size();
        int cols = rows > 0 ? grid.get(0).length() : 0;

        // Group antenna positions by frequency
        var antennas = new HashMap<Character, List<Point>>();
        for (int r = 0; r < rows; r++) {
            var row = grid.get(r);
            for (int c = 0; c < row.length(); c++) {
                char ch = row.charAt(c);
                if (ch != '.') {
                    antennas.computeIfAbsent(ch, k -> new ArrayList<>()).add(new Point(r, c));
                }
            }
        }

        return new Input(rows, cols, antennas);
    }

    /**
     * Solves Part 1: counts unique antinode positions created by pairs of antennas.
     * An antinode occurs at any point that is twice as far from one antenna as the other.
     *
     * @param input the parsed input containing grid dimensions and antenna positions
     * @return the count of unique antinode positions within the grid bounds
     */
    static int part1(Input input) {
        var antinodes = new HashSet<Point>();

        for (var entry : input.antennas.entrySet()) {
            var positions = entry.getValue();

            // For each pair of antennas with same frequency
            for (int i = 0; i < positions.size(); i++) {
                for (int j = i + 1; j < positions.size(); j++) {
                    var p1 = positions.get(i);
                    var p2 = positions.get(j);

                    // Calculate the two antinodes
                    // Antinode beyond antenna 1 (away from antenna 2)
                    int ar1 = 2 * p1.r - p2.r;
                    int ac1 = 2 * p1.c - p2.c;
                    // Antinode beyond antenna 2 (away from antenna 1)
                    int ar2 = 2 * p2.r - p1.r;
                    int ac2 = 2 * p2.c - p1.c;

                    // Add if within bounds
                    if (inBounds(ar1, ac1, input.rows, input.cols)) {
                        antinodes.add(new Point(ar1, ac1));
                    }
                    if (inBounds(ar2, ac2, input.rows, input.cols)) {
                        antinodes.add(new Point(ar2, ac2));
                    }
                }
            }
        }

        return antinodes.size();
    }

    /**
     * Solves Part 2: counts unique antinode positions with resonant harmonics.
     * Antinodes occur at any grid position exactly in line with at least two antennas
     * of the same frequency, regardless of distance.
     *
     * @param input the parsed input containing grid dimensions and antenna positions
     * @return the count of unique antinode positions within the grid bounds
     */
    static int part2(Input input) {
        var antinodes = new HashSet<Point>();

        for (var entry : input.antennas.entrySet()) {
            var positions = entry.getValue();

            // For each pair of antennas with same frequency
            for (int i = 0; i < positions.size(); i++) {
                for (int j = i + 1; j < positions.size(); j++) {
                    var p1 = positions.get(i);
                    var p2 = positions.get(j);

                    int dr = p2.r - p1.r;
                    int dc = p2.c - p1.c;

                    // Extend in both directions along the line
                    // Direction 1: from antenna 1 towards and beyond antenna 2
                    int r = p1.r;
                    int c = p1.c;
                    while (inBounds(r, c, input.rows, input.cols)) {
                        antinodes.add(new Point(r, c));
                        r += dr;
                        c += dc;
                    }

                    // Direction 2: from antenna 1 away from antenna 2
                    r = p1.r - dr;
                    c = p1.c - dc;
                    while (inBounds(r, c, input.rows, input.cols)) {
                        antinodes.add(new Point(r, c));
                        r -= dr;
                        c -= dc;
                    }
                }
            }
        }

        return antinodes.size();
    }

    /**
     * Main entry point. Parses input once and runs both parts.
     *
     * @param args command-line arguments (unused)
     */
    public static void main(String[] args) {
        try {
            var input = parseInput("../input.txt");
            System.out.println("Part 1: " + part1(input));
            System.out.println("Part 2: " + part2(input));
        } catch (IOException e) {
            System.err.println("Error reading input file: " + e.getMessage());
            e.printStackTrace();
        }
    }
}
